### DISCUSSION TITLE
Getting an error while running burnDsc test.

### DISCUSSION THREAD

**Discussion by hard1k7724** [Cyfrin/foundry-full-course-cu#1839] [Q&A] [answered]
I am getting this error when running a burnDsc test. Error ->Ran 1 test for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: panic: arithmetic underflow or overflow (0x11)] testCanBurnDsc() (gas: 42845)
Traces:
  [9684416] DSCEngineTest::setUp()
    ├─ [4371749] → new <unknown>@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 21723 bytes of code
    ├─ [5107399] 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f::run()
    │   ├─ [3336984] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372455] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [516543] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f, value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [372455] → new <unknown>@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [516543] → new <unknown>@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f, value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 6345 bytes of code
    │   ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← [Return] 
    │   ├─ [591465] → new <unknown>@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← [Return] 2610 bytes of code
    │   ├─ [938555] → new <unknown>@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← [Return] 4010 bytes of code
    │   ├─ [2424] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferOwnership(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x5FbDB2315678afecb367f032d93F642f64180aa3, 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 0x104fBc016F4bb334D775a19E8A6510109AC63E00
    ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [24912] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::mint(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], value: 10000000000000000000 [1e19])
    │   └─ ← [Stop] 
    └─ ← [Stop] 

  [42845] DSCEngineTest::testCanBurnDsc()
    ├─ [0] VM::startPrank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [24762] 0x5FbDB2315678afecb367f032d93F642f64180aa3::approve(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 100000000000000000000 [1e20])
    │   ├─ emit Approval(owner: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], spender: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, value: 100000000000000000000 [1e20])
    │   └─ ← [Return] true
    ├─ [2675] 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512::burnDsc(100000000000000000000 [1e20])
    │   └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)
    └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)
    
   test code:
   ```
   function testCanBurnDsc() public {
        vm.startPrank(USER);
        dsc.approve(address(dsce), AMOUNT_TO_MINT);
        dsce.burnDsc(AMOUNT_TO_MINT);
        vm.stopPrank();

        uint256 userBalance = dsc.balanceOf(USER);
        assertEq(userBalance, 0);
    }
    ```

**Comment by urosognjenovic** [comment]
You need to deposit collateral and mint DSC before calling the burn function. Otherwise you'll get the underflow error because you're trying to burn the amount that has never been minted.

Refer to this code for more details: https://github.com/Cyfrin/foundry-defi-stablecoin-f23/blob/main/test/unit/DSCEngineTest.t.sol.

**Reply by hard1k7724** [reply]
@thebraveman17 hey thanks, but when I run the test with the deposit and mintDSC modifier a different error is showing :
[FAIL. Reason: panic: division or modulo by zero (0x12)] testCanBurnDsc() (gas: 285564)
Traces:
  [9684416] DSCEngineTest::setUp()
    ├─ [4371749] → new <unknown>@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 21723 bytes of code
    ├─ [5107399] 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f::run()
    │   ├─ [3336984] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372455] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [516543] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f, value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [372455] → new <unknown>@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [516543] → new <unknown>@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f, value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 6345 bytes of code
    │   ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← [Return] 
    │   ├─ [591465] → new <unknown>@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← [Return] 2610 bytes of code
    │   ├─ [938555] → new <unknown>@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← [Return] 4010 bytes of code
    │   ├─ [2424] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferOwnership(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512)
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x5FbDB2315678afecb367f032d93F642f64180aa3, 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 0x104fBc016F4bb334D775a19E8A6510109AC63E00
    ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [24912] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::mint(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], value: 10000000000000000000 [1e19])
    │   └─ ← [Stop] 
    └─ ← [Stop] 

  [255264] DSCEngineTest::testCanBurnDsc()
    ├─ [0] VM::startPrank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [24739] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::approve(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 10000000000000000000 [1e19])
    │   ├─ emit Approval(owner: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], spender: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, value: 10000000000000000000 [1e19])
    │   └─ ← [Return] true
    ├─ [178785] 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512::depositCollateralAndMintDSC(0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 10000000000000000000 [1e19], 100000000000000000000 [1e20])
    │   ├─ emit CollateralDeposited(user: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], token: 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, amount: 10000000000000000000 [1e19])
    │   ├─ [30884] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::transferFrom(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 10000000000000000000 [1e19])
    │   │   ├─ emit Transfer(from: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], to: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, value: 10000000000000000000 [1e19])
    │   │   └─ ← [Return] true
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [8993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   ├─ [49149] 0x5FbDB2315678afecb367f032d93F642f64180aa3::mint(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 100000000000000000000 [1e20])
    │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], value: 100000000000000000000 [1e20])
    │   │   └─ ← [Return] true
    │   └─ ← [Stop] 
    ├─ [0] VM::stopPrank()
    │   └─ ← [Return] 
    ├─ [0] VM::startPrank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [24762] 0x5FbDB2315678afecb367f032d93F642f64180aa3::approve(0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 100000000000000000000 [1e20])
    │   ├─ emit Approval(owner: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], spender: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, value: 100000000000000000000 [1e20])
    │   └─ ← [Return] true
    ├─ [37673] 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512::burnDsc(100000000000000000000 [1e20])
    │   ├─ [26018] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferFrom(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, 100000000000000000000 [1e20])
    │   │   ├─ emit Transfer(from: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], to: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, value: 100000000000000000000 [1e20])
    │   │   └─ ← [Return] true
    │   ├─ [3422] 0x5FbDB2315678afecb367f032d93F642f64180aa3::burn(100000000000000000000 [1e20])
    │   │   ├─ emit Transfer(from: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512, to: 0x0000000000000000000000000000000000000000, value: 100000000000000000000 [1e20])
    │   │   └─ ← [Stop] 
    │   ├─ [993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← [Revert] panic: division or modulo by zero (0x12)
    └─ ← [Revert] panic: division or modulo by zero (0x12)

**Reply by urosognjenovic** [reply]
I am getting the same error on my end too.

I think the cause of the problem is the function _healthFactor.
```
function _healthFactor(address user) private view returns (uint256) {
        (uint256 totalDSCMinted, uint256 collateralValueInUSD) = _getAccountInformation(user);
        uint256 collateralAdjustedForThreshold = (collateralValueInUSD * LIQUIDATION_THRESHOLD) / 100;
        return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;
    }
```
This line of code:
```
return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;
```
breaks when the user hasn't minted any DSC, hence ```totalDSCMinted``` is zero and that's where the division with zero occurs.

I am a bit busy now so I can't help you too much, but later on during the day I'll get back to the issue.

**Reply by urosognjenovic** [reply]
Maybe just editing an if statement to avoid division with 0 zero will solve the problem? I'm not sure what to return when the user hasn't minted any DSC though. Maybe some large number?

**Reply by hard1k7724** [reply]
Same, As I am somewhat weak at maths, lol. I kinda have no idea because I'm having a tough time in understanding the math in these functions.:(( Let's see if someone else knows the solution this problem.

**Reply by hard1k7724** [reply]
Hey, @thebraveman17, I just ran a redeemCollateral test and I am getting a same modulo error in that too.

**Reply by urosognjenovic** [reply, marked as answer]
I modified the function ```_healthFactor``` to this:
```
function _healthFactor(address user) private view returns (uint256) {
        (uint256 totalDSCMinted, uint256 collateralValueInUSD) = _getAccountInformation(user);
        uint256 collateralAdjustedForThreshold = (collateralValueInUSD * LIQUIDATION_THRESHOLD) / 100;
        if (totalDSCMinted !=0) {
            return (collateralAdjustedForThreshold * PRECISION) / totalDSCMinted;
        }
        return 1e18;
    }
```

I didn't put a lot of thinking in the second return value (when ```totalDSCMinted``` is equal to zero). I just wanted a piece of code that will make my tests successful. I will eventually go back to it and figure out the best approach and then come back to this post to update you.

**Reply by hard1k7724** [reply]
sure. Thanks bro.

**Comment by EngrPips** [comment]
@Juice-07, I believe you figured this out already.

**Reply by hard1k7724** [reply]
@EngrPips, Yes kind of. as @thebraveman17 said the problem was in the _healthFactor function. As there was no (0)DSC minting in the function, the division by 0 was taking place I guess.
But I also realized that I'm pretty bad at writing these tests as I was having a really hard time and also the frustration sometimes yk. 
Though I hope I get atleast a little better. ;))

**Reply by EngrPips** [reply]
Yeah I understand the struggle but trust me, everything will make more sense with time. I use to have a very hard writing test too myself but with time and constant practice it got more clear

**Reply by hard1k7724** [reply]
Thanks, I hope It does for me too.

**Reply by hard1k7724** [reply]
Hii, @EngrPips. I have a doubt in a test I saw in the GitHub repository, I asked chatGPT but quiet didnt get it so can you please help me with that?

**Reply by EngrPips** [reply]
@Juice-07, Yeah, sure, bring it here. Let's discuss. I am sorry if I am replying slowly. It is a festive period for all Muslims worldwide, and I have been spending time with my family. I am heading back to my place today and might be slightly unstable. I will make sure to jump back on here as soon as I am stable.

**Reply by hard1k7724** [reply]
@EngrPips Please, its okay its really no problem plus you've really helped a lot of times Thanks for that. Enjoy the festival, I hope you have a really good time ;))

**Reply by hard1k7724** [reply]
So here's the test:
 function testLiquidationPayoutIsCorrect() public liquidated {
        uint256 liquidatorWethBalance = ERC20Mock(weth).balanceOf(liquidator);
        uint256 expectedWeth = dsce.getTokenAmountFromUsd(weth, amountToMint)
            + (dsce.getTokenAmountFromUsd(weth, amountToMint) / dsce.getLiquidationBonus());
        uint256 hardCodedExpected = 6_111_111_111_111_111_110;
        assertEq(liquidatorWethBalance, hardCodedExpected);
        assertEq(liquidatorWethBalance, expectedWeth);
    }
    I dont quite understand whats happening in this code, especially this "uint256 hardCodedExpected = 6_111_111_111_111_111_110;" part.

**Reply by EngrPips** [reply]
```js
function testLiquidationPayoutIsCorrect() public liquidated {
uint256 liquidatorWethBalance = ERC20Mock(weth).balanceOf(liquidator);
uint256 expectedWeth = dsce.getTokenAmountFromUsd(weth, amountToMint)
+ (dsce.getTokenAmountFromUsd(weth, amountToMint) / dsce.getLiquidationBonus());
uint256 hardCodedExpected = 6_111_111_111_111_111_110;
assertEq(liquidatorWethBalance, hardCodedExpected);
assertEq(liquidatorWethBalance, expectedWeth);
}
```
In the above test, a user (liquidator) is liquidating another user whose health_factor is below the minimum acceptable health_factor, So we are trying to confirm that the liquidator gets the appropriate `weth` they are supposed to get for successfully liquidating the user, The hard coded value must have been gotten from a computation/calculation that has been done before writing the test.
